Improved styles, made the package user-friendly, and added RbfIntegrator

Daniel-Lin-S opened the pull request:
### `container.py`
added docstrings and comments
added mins and maxs handling to allow arbitrary boundaries in each dimension
added filtering of points outside the container using `filter_points` method
modified `rvs` to support infinite boundaries.
changed Assertions to ValueErrors with clear error messages

### `containerIntegration.py`
added docstrings to existing methods
added `rbfIntegral`
added an abstract base class `ContainerIntegral` and modified existing methods to a class
- easier to specify integrator parameters in this way
Split it into separate files in `containerIntegration`, similar to `splits` and `integrators`

### added `gaussianProcess.py`
with methods necessary to fit GP to a container

### `visualisation.py`
added `plotContainers` that automatically plots all containers
1-dimensional problems are also also supported by `_plotContainers1D`. 
added `plotIntegrand` to visualise the integrand
added `plotGP` which plots the posterior mean in 1-D cases.  
edited `_plot2DContainers` method so that it supports plotting of any two dimensions for higher-dimensional problems.

### `exampleDistributions.py`
added docstrings
`MixtureDistribution` changed to handle arbitrary weights 
created abstract class Distribution to allow User to define distribution
fixed `Uniform.pdf`: 
- now will not return all 0 even if only one value is outside the range
- and properly scales as the dimension increases by fixing `low` and `high` using `_handle_low_high` method.

added a test in `tests/test_distributions.py` to ensure Uniform.pdf returns correct values and Uniform.rvs samples are within bounds

### `exampleProblems.py`
added docstrings
Added checks to `Problem.pdf` on the shape of X
changed original `Problem` to `BayesProblem` 
created a class `Problem` for arbitrary problems with only an `integrand` 
added `Gaussian` that represents integrating the pdf of any Multivariate Gaussian distribution on given boundaries
added `PyramidProblem` that does not have `rvs` method to test samplers. 

### added `sampler`, `uniformSampler.py`, `importanceSampler.py` , `mcmcSampler.py` 
to accommodate for arbitrary problems without method `rvs` 
modified `TreeIntegratror` so that when `problem` does not have `rvs`,  `self.sampler` will be used. (defaults to UniformSampler) 
the corresponding tests can be found in `tests/test_samplers.py`

### `utils.py`
added docstring
added exception handling to `scale` (prevent scaling constant array) 

### added `integrators/integrator.py`
an abstract base class `Integrator`  for all other integrators 
test_integrators has been modified to accomodate the changes

### added `integrators/treeIntegrator.py`
an abstract base class inheriting `Integrator`, represents a tree-based integrator
- must have `constrct_tree` method that constructs a list of containers from a root container
- `__call__` method is the same for all trees: aggregate the results in the trees.

modified `__call__` to allow returning standard deviation estimation when containerIntegration method has return_std option. 
modified `__call__` to return a dictionary, which is easier to use. 

modified the tests correspondingly

### added `integrators/bayesMCIntegrator.py` 
performs Bayes Monte Carlo 

and added to `tests/test_integrators.py`

### added `gpTreeIntegrator.py` 
a more powerful GP fitting scheme with parameter passing between containers. 

and added to `tests/test_integrators.py`

### `integrators/smcIntegrator.py`, `integrators/vegasIntegrator.py`, `integrators/limitedSampleIntegrator.py`, `integrators/queueIntegrator.py`, `integrators/simpleIntegrator.py`
added docstrings with examples for users 
modified to inherit from `Integrator` 
added better type handling (avoid Users giving wrong types as arguments) 
`LimitedSampleIntegrator`, `QueueIntegrator` and `SimpleIntegrator` now all inherit from `TreeIntegrator`
added `verbose` option to `TreeIntegrator` for easier debugging and tracking.  

### added `splits/split.py`
an abstract base class `Split` for all split methods for container

### `splits/uniformSplit.py`
implemented `uniformSplit`

### `splits/kdSplit.py`, `splits/minSseSplit.py`
added docstrings
modified to inherit from `Split` 
added a check for X with less than 2 unique values. (cannot split anymore) in `KdSplit`

### `example_multivariate_normal.ipynb`
import containerIntegrals from the package instead so that Users know how to use this package 

### `tests/test_integrators.py`
added test_io for all integrators
added UniformSplit and RbfIntegral to the tests

### added `compare_integrators.py`
conveniently tests a set of integrators and problems, and writes the results into a csv file.

Daniel-Lin-S committed:
Added docstrings, added rbfIntegrator to container
Integration

Daniel-Lin-S committed:
cleaned the codes in simpleIntegrator

Daniel-Lin-S committed:
Cleaned unnecessary argument in rbfIntegral

Daniel-Lin-S committed:
Update requirements.txt

Daniel-Lin-S committed:
added methods to plot integrand and GP

Daniel-Lin-S committed:
Merge branch 'gaussian-process' of https://github.com/thomfoster/treeQuadrature into gaussian-process

Daniel-Lin-S committed:
changed import

Daniel-Lin-S committed:
addded docstrings to three integrators and utils.

Daniel-Lin-S committed:
Update container.py

changed np.product to np.prod

Daniel-Lin-S committed:
optimised exampleProblems and exampleDistributions
added abstract class Distribution, added checks to Problem.pdf,
fixed Uniform.pdf which should NOT return all zeros when only one value is out of range.
Changed MixtureDistribution to allow arbitrary weights.

Daniel-Lin-S committed:
fixed the dimension of X

Daniel-Lin-S committed:
Implemented uniformSplit, modified kdSplit

Daniel-Lin-S committed:
improved functionality of Container
added `filter_points` to filter out points not in the container
added docstrings
modified rvs to handle infinite bounds

Daniel-Lin-S committed:
removed draft codes

Daniel-Lin-S committed:
Fixed Uniform.pdf and added test_uniform

Daniel-Lin-S committed:
fixed typo in visualisation.py

Daniel-Lin-S committed:
added a test for containers with infinite volume.

Daniel-Lin-S committed:
Modified Problem and add general Gaussian problem
allow flexible bounds in Problem

Daniel-Lin-S committed:
Updated example_multivariate_normal
implemented using methods from the package, easier for users to understand

Daniel-Lin-S committed:
added proper error handling

Daniel-Lin-S committed:
Wrapped all splits into an abstract class Split

Daniel-Lin-S committed:
added type handling of Distribution

Daniel-Lin-S committed:
wrapped all containerIntegrators to an asbtract class

Daniel-Lin-S committed:
Changed all integrators correspondingly
added user examples

Daniel-Lin-S committed:
modified test_integrators correspondingly
added RbfIntegral to the test.

Daniel-Lin-S committed:
changed AssertionError to ValueError

Daniel-Lin-S committed:
Created an abstract class Integrator

Daniel-Lin-S committed:
avoid errors caused by int input to Gaussian

Daniel-Lin-S committed:
split containerIntegration into separate files

Daniel-Lin-S committed:
added a test on constructing non-square containers

Daniel-Lin-S committed:
Modified all integrators to return a dict
easier to access the results and keep track

Daniel-Lin-S committed:
Wrapped all Integrators relating to trees in an abstract class
see treeIntegrator

Daniel-Lin-S committed:
Added UniformSplit and IO checks to test_integrators

Daniel-Lin-S committed:
modified example_multivariate_normal accordingly

Daniel-Lin-S committed:
Fixed the docstrings of Integrators.

Daniel-Lin-S committed:
gave names to the integrators, fixed return_std

Daniel-Lin-S committed:
added return_std option to SmcIntegrator

Daniel-Lin-S committed:
Created a new abstract class Problem
accomodates any integration problem, not just Bayesian with p and d

Daniel-Lin-S committed:
moved handle_bound to utils

Daniel-Lin-S committed:
added _handle_X_y to container
if y has shape (N,), will be converted to (N, 1) automatically

Daniel-Lin-S committed:
added keywords to test_integrators

Daniel-Lin-S committed:
added samplers to accomodate for Problem
problem may not have rvs method.

Daniel-Lin-S committed:
Improved type handling

Daniel-Lin-S committed:
added convenient compare_integrators method
for user testing and develouper testing.

Daniel-Lin-S committed:
Added McmcSampler and test_samplers

Daniel-Lin-S committed:
Added verbose option for debugging and tracking

Daniel-Lin-S committed:
added PyramidProblem

Daniel-Lin-S committed:
Adjusted visualisation to allow higher dimensions

Daniel-Lin-S committed:
Adjusted typing and docstrings

Daniel-Lin-S committed:
changed RbfIntegral to return length_scale (opt)

Daniel-Lin-S committed:
adde BMC Integrator and a corresponding test

Daniel-Lin-S committed:
fixed problems in RbfIntegral

Daniel-Lin-S committed:
added _plotGP2D for 2D problems

Daniel-Lin-S committed:
added iterative GP fitting scheme

Daniel-Lin-S committed:
added GpTreeIntegrator

Daniel-Lin-S committed:
Fixed problems in last commit

Daniel-Lin-S committed:
fixed f-string

Daniel-Lin-S committed:
give names to problems.

Daniel-Lin-S committed:
added test_integrator that stores output into csv

Daniel-Lin-S committed:
modified test_integrators to allow multiple runs

Daniel-Lin-S committed:
created GPFit base class for user-defined GP fitter

Daniel-Lin-S committed:
modified BayesMcIntegrator accordingly

Daniel-Lin-S committed:
modified typing of SmcIntegrator to BayesProblem

Daniel-Lin-S committed:
modified rbf_integration to kernel_integration
so that other kernels can be used by user

Daniel-Lin-S committed:
modified related integrators correspondingly

Daniel-Lin-S committed:
modified minSseSplit to dynamic programming
efficiency enhanced.

Daniel-Lin-S committed:
added parallel computing to TreeIntegrator

Daniel-Lin-S committed:
moved min_samples_leaf to __init__

Daniel-Lin-S committed:
merged `compare_integrators` from backup branch

Daniel-Lin-S committed:
modified IterativeGPFit to allow parallel compute

Daniel-Lin-S committed:
store iGP and kernel as attribute for RbfIntegral

Daniel-Lin-S committed:
modified GpTreeIntegrator to allow batch fitting
and parallel computation

Daniel-Lin-S committed:
modified test_integrators to store attributes

Daniel-Lin-S committed:
Merged changes from gp-backup-branch

Daniel-Lin-S committed:
updated compare_integrators docstrings

Daniel-Lin-S committed:
Changed absolute errors to relative errors

Daniel-Lin-S committed:
added functions to plot csv files
and changed compare_integrators to support multiple runs

Daniel-Lin-S committed:
added pandas for previous commit

Daniel-Lin-S committed:
fixed bugs

Daniel-Lin-S committed:
Merged Changes from gp-backup-branch
Now gp supports fitting to mean

Daniel-Lin-S committed:
Fixed problems in previous commit

Daniel-Lin-S committed:
Improved numerical stability

Daniel-Lin-S committed:
Optimised return logic for GP

Daniel-Lin-S committed:
fixed problems in plot_errors and plot_times

Daniel-Lin-S committed:
added simple problems with exact solutions

otto-ap committed:
fixed MinSseSplit sum_left

otto-ap committed:
Merge branch 'gaussian-process' of https://github.com/thomfoster/treeQuadrature into local-gp

otto-ap committed:
Fixed sum_left in MinSseSplit

otto-ap committed:
Merge branch 'gaussian-process' of https://github.com/thomfoster/treeQuadrature into gaussian-process

Daniel-Lin-S committed:
Merged changes from gp-backup-branch

Daniel-Lin-S committed:
Fixed problems in plot_errors

Daniel-Lin-S committed:
Merged changes from backup branch.

Daniel-Lin-S committed:
changed test_integrators correspondingly

Daniel-Lin-S committed:
Fixed problems in IS sampler in last version

Daniel-Lin-S committed:
Split gaussianProcess into separate files

otto-ap committed:
added problems

otto-ap committed:
added integrators

otto-ap committed:
Merge branch 'gaussian-process' of https://github.com/thomfoster/treeQuadrature into local-gp

Daniel-Lin-S committed:
added default scoring

Daniel-Lin-S committed:
Removed incomplete integrators.

Daniel-Lin-S committed:
Fixed shapes in plot_1D and problems

Daniel-Lin-S committed:
fixed typing

Daniel-Lin-S committed:
fixed typing

Daniel-Lin-S committed:
Added LimitedSampleGpIntegrator

Daniel-Lin-S committed:
Added two samplers that aids higher dimensions

Daniel-Lin-S committed:
fixed importanceSampler

Daniel-Lin-S committed:
fixed samplers and added the tests

Daniel-Lin-S committed:
Merged backup branch changes on samplers

Daniel-Lin-S committed:
added docstrings and fixed None type handling

Daniel-Lin-S committed:
Added an integrator that controls number of samples
used

Daniel-Lin-S committed:
Fixed attribute problem

Daniel-Lin-S committed:
Merged Changes from GP backup branch

Daniel-Lin-S committed:
Added new samplers to the tests

thomfoster committed:
add install_requires to setup.py to install requirements during install

thomfoster commented on treeQuadrature/container.py:
Do we need this capacity thing? Does it save that much time doing dynamic allocation of memory?

Daniel-Lin-S replied:
Are you suggesting setting the container's capacity during initialization based on the size of the initial samples?
Or should we simply use numpy.stack to reallocate a new array each time?

thomfoster replied:
I think we can just do numpy.stack :)

thomfoster commented on treeQuadrature/container.py:
```diff
@@ -38,44 +39,167 @@ def contents(self):
+        """
+
+        if np.any(mins == np.inf):
+            raise ValueError(f'mins cannot have np.inf, got {mins}')
```
Why add this? What if I want to specify a container that is bounded only on 1 side for example?

Daniel-Lin-S replied:
Then one would define e.g. mins = [-np.inf, 1] and maxs = [np.inf, 3]. But no entry of mins can be np.inf. In this case, no point can be added and the programme will collapse.

thomfoster commented on treeQuadrature/container.py:
```diff
@@ -38,44 +39,167 @@ def contents(self):
+        self.mins = self._handle_min_max_bounds(mins, -np.inf) 
+        self.maxs = self._handle_min_max_bounds(maxs, np.inf)
+        # dimensionality checks
+        if self.mins.shape[0] != self.D:
```
Move shape checks inside _handle_min_max_bounds

thomfoster commented on treeQuadrature/container.py:
```diff
@@ -38,44 +39,167 @@ def contents(self):
+        new_X, new_y = self._handle_X_y(new_X, new_y)
+        
+        if not np.all(new_X >= self.mins):
+            raise ValueError(f"Some values in new_X are below the minimum bounds: {new_X[new_X < self.mins]}")
```
Why do this when we've already thrown a warning?

thomfoster commented on treeQuadrature/container.py:
```diff
@@ -92,14 +216,43 @@ def X(self):
-        rs = np.random.uniform(size=(n, self.D))
-        ranges = self.maxs - self.mins
-        rs = self.mins + ranges * rs
+    def rvs(self, n: int):
```
nice!

thomfoster commented on tests/test_container.py:
Nice job adding these tests

thomfoster commented on treeQuadrature/containerIntegration/containerIntegral.py:
I don't love how we've done this - I'd quite like to refactor this such that ContainerIntegrals aren't classes, they are just functions that operate on Containers and return an integral number + results dict. What do you think?

Daniel-Lin-S replied:
Changing these classes to functions is possible for simple constant container integrals but unsuitable for GP integrals, which are more complicated to manage.

thomfoster commented on treeQuadrature/containerIntegration/constantIntegral.py:
```diff
@@ -0,0 +1,45 @@
+        self.name = 'MedianIntegral'
+
+    def containerIntegral(self, container: Container, f: callable, 
+                          return_std: bool=False):
```
return_std breaks that containerIntegral's should always return a dict with a key of 'integral' that is a single float value. Should instead return {'integral': float, 'std': float} for the case where n==0

thomfoster committed:
Fix niche bug for median integral when return_std=True and container.N == 0

thomfoster commented on treeQuadrature/containerIntegration/monteCarloIntegral.py:
nice work

thomfoster commented on treeQuadrature/gaussianProcess.py:
This should be deleted right? in place of `/gaussianProcess/gaussianProcess.py`

thomfoster commented on treeQuadrature/exampleProblems.py:
```diff
@@ -1,77 +1,962 @@
+        dimension of the problem
+    lows, highs : np.ndarray
+        the lower and upper bound of integration domain
+        assumed to be the same for each dimension
```
this is a crazy thing to restrict to but I guess we're doing it ahaha

thomfoster commented on treeQuadrature/exampleDistributions.py:
I think this is confusing and maybe obselete now - do you agree? we can just merge this with exampleProblems and clean up?

thomfoster commented on treeQuadrature/splits/uniformSplit.py:
```diff
@@ -1,4 +1,49 @@
+
+        subcontainers = []
+
+        # Iterate over all possible combinations of split points
```
If this is slow and we want to speed it up I think we can remove the for loops and do as straight numpy

thomfoster commented on treeQuadrature/samplers/sobolSampler.py:
cool! Hadn't heard of this before

thomfoster commented on treeQuadrature/samplers/mcmcSampler.py:
Nice! Wild that you implemented this yourself - i would have definitely just used a library so I learnt something here!

thomfoster commented on treeQuadrature/samplers/lowDiscrepancySampler.py:
Replication of sobolSampler.py??

thomfoster commented on treeQuadrature/integrators/vegasIntegrator.py:
```diff
@@ -9,24 +11,45 @@ def __call__(self, X):
+        """
         integ = vegas.Integrator([[-1.0, 1.0]] * problem.D)
-        f = ShapeAdapter(problem.pdf)
+        f = ShapeAdapter(problem.integrand)
```
Nice I like this change to using integrand

thomfoster commented on treeQuadrature/integrators/limitedSampleIntegrator.py:
```diff
@@ -26,72 +29,159 @@ def save_weights_image(q):
 
-class LimitedSampleIntegrator:
-
+class LimitedSampleIntegrator(TreeIntegrator):
```
Should this inherit from QueueIntegrator?

thomfoster commented on treeQuadrature/integrators/limitedSampleGpIntegrator.py:
Isn't most of this code already written in limited sample integrator?

thomfoster commented on treeQuadrature/integrators/integrator.py:
A nice addition here would be typing the return value of the integrators, ie returning a TypedDict or dataclass that we specify must contain the 'estimate' key

thomfoster commented on treeQuadrature/integrators/gpTreeIntegrator.py:
As with the limited sample integrator, let's clean this up and make use of code sharing

thomfoster commented on treeQuadrature/gaussianProcess/visualisation.py:
nice

Daniel-Lin-S committed:
git c

Daniel-Lin-S committed:
Fixed problems in last commit
allow user control of growth rate and initial capacity in ArrayList

Daniel-Lin-S committed:
fixed GpTreeIntegrator return_std

Daniel-Lin-S committed:
Used ensemble MCMC sampler instead of original one
and changed SobolSampler to generate exact number of samples

Daniel-Lin-S committed:
added option to keep initial samples in gpIntegral

Daniel-Lin-S committed:
enhanced sample distribution to bias towards
larger containers

Daniel-Lin-S committed:
fixed Vegas integrator

Daniel-Lin-S committed:
fixed other minor problems

Daniel-Lin-S committed:
added emcee package for mcmc sampler

Daniel-Lin-S committed:
SMC should only take Bayes problem

Daniel-Lin-S committed:
Made minSseSplit more flexible

Daniel-Lin-S committed:
added mixed sampler and more docstrings to problems

thomfoster reviewed (approved):
This is great work Daniel. Happy to merge this into main.

Moving forward, we need to:
- Add a top level readme that
    - explains the key absract base classes
    - provides a snippet of code that installs the library and runs the tests
    - provides a snippet of code to integrate your own problem
- Produce a notebook for visualisation of the problems
- Clean up code sharing between the various integrator classes

Daniel-Lin-S committed:
Resolved problems mentioned in conversations

Daniel-Lin-S committed:
Fixed problems in previous commit

Daniel-Lin-S committed:
Fixed docstrings and modified README.md

Daniel-Lin-S committed:
Added documentation

Daniel-Lin-S committed:
Fixed styling issues.

Daniel-Lin-S committed:
Fixed flake8 styles

Daniel-Lin-S committed:
new integrator (VegasTree) and other minor changes

Daniel-Lin-S committed:
Fixed problem in last commit

Daniel-Lin-S commented:
Dear Thom,
Thank you for your feedback on the ongoing projects. I have addressed all the points you mentioned. Regarding the notebook of problems, I believe Otto has already prepared one.
Please let me know if there is anything else that needs attention or if further modifications are required.
Best regards,
Daniel

From: thomfoster ***@***.***>
Date: Tuesday, 27 August 2024 at 14:14
To: thomfoster/treeQuadrature ***@***.***>
Cc: Daniel Lin ***@***.***>, Author ***@***.***>
Subject: Re: [thomfoster/treeQuadrature] Improved styles, made the package user-friendly, and added RbfIntegrator (PR #27)

@thomfoster approved this pull request.

This is great work Daniel. Happy to merge this into main.

Moving forward, we need to:

  *   Add a top level readme that
     *   explains the key absract base classes
     *   provides a snippet of code that installs the library and runs the tests
     *   provides a snippet of code to integrate your own problem
  *   Produce a notebook for visualisation of the problems
  *   Clean up code sharing between the various integrator classes

________________________________

In treeQuadrature/container.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722186623>:

> -        assert y.shape[1] == 1

+    def __init__(self, X: np.ndarray, y: np.ndarray, mins=None, maxs=None):

+        """

+        Attributes

+        ----------

+        X : numpy.ndarray of shape (N, D)

+            each row is a sample

+        y : numpy.ndarray of shape (N, 1) or (N,)

+            the function value at each sample

+        mins, maxs : int or float or list or numpy.ndarray of shape (D,), optional

+            the low and high boundaries of the hyper-rectangle

+            could be +- np.inf

+        """

+

+        if np.any(mins == np.inf):

+            raise ValueError(f'mins cannot have np.inf, got {mins}')

Why add this? What if I want to specify a container that is bounded only on 1 side for example?

________________________________

In treeQuadrature/container.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722187311>:

>

         self.D = X.shape[1]



-        # Compute container properties

-        self.mins = np.array(mins) if mins is not None else np.array(

-            [-np.inf] * self.D)

-        self.maxs = np.array(maxs) if maxs is not None else np.array(

-            [np.inf] * self.D)

-        self.volume = np.product(self.maxs - self.mins)

+        # if mins (maxs) are None, create unbounded container

+        self.mins = self._handle_min_max_bounds(mins, -np.inf)

+        self.maxs = self._handle_min_max_bounds(maxs, np.inf)

+        # dimensionality checks

+        if self.mins.shape[0] != self.D:

Move shape checks inside _handle_min_max_bounds

________________________________

In treeQuadrature/container.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722188757>:

>

     def add(self, new_X, new_y):

-        assert new_X.ndim == 2

-        assert new_y.ndim == 2

-        assert new_X.shape[0] == new_y.shape[0]

-        assert new_X.shape[1] == self.D

-        assert new_y.shape[1] == 1

-        assert np.all(new_X >= self.mins), new_X[new_X < self.mins]

-        assert np.all(new_X <= self.maxs), new_X[new_X > self.maxs]

+        new_X, new_y = self._handle_X_y(new_X, new_y)

+

+        if not np.all(new_X >= self.mins):

+            raise ValueError(f"Some values in new_X are below the minimum bounds: {new_X[new_X < self.mins]}")

Why do this when we've already thrown a warning?

________________________________

In treeQuadrature/container.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722189899>:

> @@ -92,14 +216,43 @@ def X(self):

     def y(self):

         return self._y.contents



-    def rvs(self, n):

-        rs = np.random.uniform(size=(n, self.D))

-        ranges = self.maxs - self.mins

-        rs = self.mins + ranges * rs

+    def rvs(self, n: int):

nice!

________________________________

On tests/test_container.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722193903>:

Nice job adding these tests

________________________________

On treeQuadrature/containerIntegration/containerIntegral.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722199765>:

I don't love how we've done this - I'd quite like to refactor this such that ContainerIntegrals aren't classes, they are just functions that operate on Containers and return an integral number + results dict. What do you think?

________________________________

In treeQuadrature/containerIntegration/constantIntegral.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722201903>:

> +    '''estimate integral by the function value at mid-point of container'''

+    def containerIntegral(self, container: Container, f: callable):

+        mid_x = container.midpoint

+        mid_y = f(mid_x)[0, 0]

+        integral_estimate = mid_y * container.volume

+

+        return {'integral' : integral_estimate}

+

+

+class MedianIntegral(ContainerIntegral):

+    '''estimate integral by the median of samples in the container. '''

+    def __init__(self):

+        self.name = 'MedianIntegral'

+

+    def containerIntegral(self, container: Container, f: callable,

+                          return_std: bool=False):

return_std breaks that containerIntegral's should always return a dict with a key of 'integral' that is a single float value. Should instead return {'integral': float, 'std': float} for the case where n==0

________________________________

On treeQuadrature/containerIntegration/monteCarloIntegral.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722215573>:

nice work

________________________________

On treeQuadrature/gaussianProcess.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722216945>:

This should be deleted right? in place of /gaussianProcess/gaussianProcess.py

________________________________

In treeQuadrature/exampleProblems.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722218017>:

>

-class Problem:

-    def __init__(self, D):

-        self.D = None

-        self.d = None

-        self.low = None

-        self.high = None

-        self.p = None

+    Attributes

+    ----------

+    D : int

+        dimension of the problem

+    lows, highs : np.ndarray

+        the lower and upper bound of integration domain

+        assumed to be the same for each dimension

this is a crazy thing to restrict to but I guess we're doing it ahaha

________________________________

On treeQuadrature/exampleDistributions.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722227163>:

I think this is confusing and maybe obselete now - do you agree? we can just merge this with exampleProblems and clean up?

________________________________

In treeQuadrature/splits/uniformSplit.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722231604>:

> +        n_divisions = kwargs.get('n_divisions', self.n_divisions)

+

+        # Ensure n_divisions is positive

+        if n_divisions <= 0:

+            raise ValueError("n_divisions must be a positive integer")

+

+        # Get the mins and maxs of the container

+        mins = container.mins

+        maxs = container.maxs

+

+        # Generate the split points along each dimension

+        split_points = [np.linspace(mins[i], maxs[i], n_divisions + 1) for i in range(container.D)]

+

+        subcontainers = []

+

+        # Iterate over all possible combinations of split points

If this is slow and we want to speed it up I think we can remove the for loops and do as straight numpy

________________________________

On treeQuadrature/samplers/sobolSampler.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722233705>:

cool! Hadn't heard of this before

________________________________

On treeQuadrature/samplers/mcmcSampler.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722237004>:

Nice! Wild that you implemented this yourself - i would have definitely just used a library so I learnt something here!

________________________________

On treeQuadrature/samplers/lowDiscrepancySampler.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722237802>:

Replication of sobolSampler.py??

________________________________

In treeQuadrature/integrators/vegasIntegrator.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722240924>:

>          integ = vegas.Integrator([[-1.0, 1.0]] * problem.D)

-        f = ShapeAdapter(problem.pdf)

+        f = ShapeAdapter(problem.integrand)

Nice I like this change to using integrand

________________________________

In treeQuadrature/integrators/limitedSampleIntegrator.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722292511>:

> @@ -26,72 +29,159 @@ def save_weights_image(q):

     plt.close()





-class LimitedSampleIntegrator:

-

+class LimitedSampleIntegrator(TreeIntegrator):

Should this inherit from QueueIntegrator?

________________________________

On treeQuadrature/integrators/limitedSampleGpIntegrator.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722293376>:

Isn't most of this code already written in limited sample integrator?

________________________________

On treeQuadrature/integrators/integrator.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722294465>:

A nice addition here would be typing the return value of the integrators, ie returning a TypedDict or dataclass that we specify must contain the 'estimate' key

________________________________

On treeQuadrature/integrators/gpTreeIntegrator.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722294959>:

As with the limited sample integrator, let's clean this up and make use of code sharing

________________________________

On treeQuadrature/gaussianProcess/visualisation.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1722296531>:

nice

________________________________

On treeQuadrature/container.py<https://github.com/thomfoster/treeQuadrature/pull/27#discussion_r1732825925>:

I think we can just do numpy.stack :)

—
Reply to this email directly, view it on GitHub<https://github.com/thomfoster/treeQuadrature/pull/27#pullrequestreview-2246193090>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASDZYQFGLRYYZ5O3OAAJLKTZTR3UZAVCNFSM6AAAAABK2MNULCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDENBWGE4TGMBZGA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Daniel-Lin-S committed:
Updated changes from backup-branch

Daniel-Lin-S committed:
fixed test parameters of VegasTreeIntegrator

Daniel-Lin-S committed:
Fix type logic in filter_points

Daniel-Lin-S committed:
changed warning settings

Daniel-Lin-S committed:
Merged changes from backup branch:
added cuba and refined vegas+TQ approach.

Daniel-Lin-S committed:
added required package.

Daniel-Lin-S committed:
removed tests on integrators from exterior packages
